### DISCUSSION TITLE
Triple Charger Controller Status

### DISCUSSION THREAD

**Discussion by sjoram** [syssi/esphome-votronic#55] [Q&A]
Hi,
   First off many thanks for this integration. It is working great. I, however have a triple charger. The data set supports the PV supply but I'd love to get information about the other two modes - B2B and grid power. I can assume them from my SmartShunt but would love to get them directly from the Votronic if possible.
I can see that the Controller Status must have some further information that is not fully decoded. For example when I plug in my grid the  status changes to "Reduced". At the moment it is night time and I am not plugged in and the status is "unknown".
Unfortunately, I don't have access to a Votronic display so I don't know what that would display under these conditions that might be attributable to the status.
Has anybody else had a look at this and cold help decode it further? Are there any other fields that might report the B2B and grid charge elements?

**Comment by syssi** [comment]
Could you enable the debug output of the `uart` component and provide some logs?

https://github.com/syssi/esphome-votronic/blob/main/votronic_display_link.md#debugging

**Reply by sjoram** [reply]
Ok many thanks for the speedy reply. I will definitely do this but it will be the end of the week before I can reflash the ESP when I get home.

**Reply by syssi** [reply]
Some RAW traffic of you setup would be helpful to get an idea of whether we can decode more.

**Reply by sjoram** [reply]
Yes. I'll try to log that at the weekend. I'll also force B2B and grid charging and associate times with them. Maybe values from other sources.

**Reply by sjoram** [reply]
OK took a lot longer than I had hoped but I have just logged the UART. The first minute or so is with no external power connected but after 16:46 I connected the external 220VAC.
I will also try to do another run with the B2B section charging but that will be a day or so!
Is the log sufficient to see anything? The battery should be charging at around 15.X amps.

[logs_votronic_run.txt](https://github.com/user-attachments/files/18912706/logs_votronic_run.txt)

**Reply by syssi** [reply]
Awesome! I've reformatted the capture manually for further analysis:

[logs_votronic_run-raw-only.txt](https://github.com/user-attachments/files/18913442/logs_votronic_run-raw-only.txt)

**Reply by syssi** [reply]
It looks like there are 4 different devices/frames on the bus:

```
# VOTRONIC_FRAME_TYPE_CONTROL_CHARGING_CONVERTER
AA:3A:FD:04:F9:04:00:00:00:01:00:9A:52:28:00:DF
AA:3A:FD:04:F8:04:00:00:00:01:00:9A:52:28:00:DE
AA:3A:FD:05:DD:07:00:00:00:E0:9A:11:52:20:11:C4
AA:3A:09:05:02:05:C8:00:00:01:63:9A:52:28:0C:77
AA:3A:09:05:02:05:C8:00:00:01:64:9A:52:28:0C:70
AA:3A:09:05:02:05:C8:00:00:01:64:9A:52:28:0C:70
```

```
# VOTRONIC_FRAME_TYPE_CONTROL_CHARGING_CONVERTER2
AA:4A:00:00:00:00:00:00:00:00:00:00:00:00:00:4A
AA:4A:00:00:00:00:00:00:00:00:00:00:00:00:00:4A
AA:4A:00:00:00:00:00:00:00:00:00:00:9A:52:28:00:9F
AA:4A:00:00:00:00:00:00:00:00:00:00:00:00:00:4A
AA:4A:00:00:00:00:00:00:00:00:00:00:00:00:00:4A
```

```
# VOTRONIC_FRAME_TYPE_SOLAR_CHARGER
AA:1A:FD:04:94:05:01:00:00:E0:9A:11:52:20:09:63
AA:1A:FD:04:94:05:01:00:00:E0:9A:11:52:20:09:63
AA:1A:FD:04:96:05:01:00:00:E0:9A:11:52:20:09:61
AA:1A:FD:04:96:05:01:00:00:E0:9A:11:52:20:09:61
AA:1A:09:05:DD:07:00:00:00:E0:9A:11:52:20:11:C4
AA:1A:09:05:DD:07:00:00:00:E0:9A:11:52:20:11:C4
AA:1A:09:05:DD:07:00:00:00:E0:9A:11:52:20:11:C4
AA:1A:09:05:DD:07:00:00:00:E0:9A:11:52:20:11:C4
AA:1A:0A:05:DD:07:00:00:00:E0:9A:12:52:20:11:C4
AA:1A:0A:05:DD:07:00:00:00:E0:9A:12:52:20:11:C4
AA:1A:0A:05:DD:07:00:00:00:E0:9A:12:52:20:11:C4
AA:1A:0A:05:DD:07:00:00:00:E0:9A:12:52:20:11:C4
```

```
# VOTRONIC_FRAME_TYPE_CHARGING_CONVERTER
AA:7A:FD:04:F9:04:00:00:00:00:00:9A:52:28:00:9E
AA:7A:FD:04:F9:04:00:00:00:00:00:9A:52:28:00:9E
AA:7A:FD:04:F8:04:00:00:00:00:00:9A:52:28:00:9F
AA:7A:FD:04:F8:04:00:00:00:00:00:9A:52:28:00:9F
AA:7A:09:05:02:05:00:00:00:00:00:9A:52:28:00:91
AA:7A:09:05:02:05:00:00:00:00:00:9A:52:28:00:91
AA:7A:09:05:02:05:00:00:00:00:00:9A:52:28:00:91
AA:7A:09:05:02:05:00:00:00:00:00:9A:52:28:00:91
AA:7A:0A:05:02:05:00:00:00:00:00:9A:52:28:00:92
AA:7A:0A:05:02:05:00:00:00:00:00:9A:52:28:00:92
AA:7A:0A:05:02:05:00:00:00:00:00:9A:52:28:00:92
AA:7A:0A:05:02:05:00:00:00:00:00:9A:52:28:00:92
AA:7A:0C:05:19:05:00:00:00:00:00:9A:52:28:00:8F
AA:7A:0C:05:19:05:00:00:00:00:00:9A:52:28:00:8F
AA:7A:0C:05:19:05:00:00:00:00:00:9A:52:28:00:8F
AA:7A:0C:05:19:05:00:00:00:00:00:9A:52:28:00:8F
```

**Reply by sjoram** [reply]
I've been doing some digging myself since. My 'Starter Battery' is visible as the 2nd Battery in the Charger Data Decode (13.09V) but this is not exposed to Home Assistant. Only the Solar Charger Data seems to be exposed.

**Reply by syssi** [reply]
Could you provide a list of all votronic devices do you run? I would like to get a better idea how many real devices are on the bus especially and the number of ways to charge your batteries. How is the starter battery and the 2nd battery connected? Do you always charge both battery simultaneously?

**Reply by sjoram** [reply]
I only have one device! The Triple charger. I am connected directly to the display port of that device and no other....
It has a solar, B2B and 220VAC charger in built and switches between them automatically as appropriate.
Both batteries are also connected to the same charger and it decides how and what to charge or maintain.

**Reply by sjoram** [reply]
A bit more digging on the Votronic site gave the attached manual for the 'LCD Charge Control S'.

What I would like to be able to do is decode the operational status as in the attached snip if possible.

![image](https://github.com/user-attachments/assets/6747babc-817b-41d9-abb3-eb65eb5cbf22)

[1247_d.pdf](https://github.com/user-attachments/files/18930413/1247_d.pdf)

**Reply by sjoram** [reply]
OK here is some more information. I pasted a bunch of screenshots and the raw uart to a Word document attached.

The 1A is the solar controller
The 3A is the AC supply controller
The 7A is the alternator / B2B

I had the solar controller on first then connected the AC then disconnected the AC and started the engine. These are the 3 segments in the document.

Bytes 6 & 7 in the 3A and 7A telegrams are the current as I can map them to the values of my Smartshunt.
The 7A is not parsed in the code but also I see there should be a current value exposed from the 3A charge controller but it is not. My programming skills are limited and there are multiple layers of abstraction that I can't follow.

I believe the 'Votronic Charging' and 'Votronic Controller Active' status bits my be incorrectly interpreted as they oscillate when the charge controller is actually active.

I also believe that the 'Votronic Controller Temperature' may be incorrectly interpreted as it is always 151C irrespective of what else is happening.

[Votronic Data 240225.docx](https://github.com/user-attachments/files/18949201/Votronic.Data.240225.docx)

**Reply by syssi** [reply]
Alright! In this case it doesn't make sense to route the measurements of 0x3A and 0x7A into the same sensor entities. I will introduce a new set of sensors to provide individual measurements per device.

**Reply by syssi** [reply]
> I also believe that the 'Votronic Controller Temperature' may be incorrectly interpreted as it is always 151C irrespective of what else is happening.

I guess it's 15.1°C.

**Reply by sjoram** [reply]
Regarding the 15.1 - that could make sense BUT it doesn't really move. However, the 'PV Controller Temperature' does - a lot!
You can see in my first segment that it is around 16C and stable under Solar production. As soon as the AC comes on it shoots up to 24C and when the alternator is on (30 Amps) then it is over 40C. And this was only a short period. The 15.1 didn't really move so I'm not sure what it is representing.

**Reply by sjoram** [reply]
Also I never see anything except 'U3 Phase' for the battery status :) - irrespective of real state of charge!

Anyway, I'm not nitpicking just trying to understand myself. I'll try to map out the bit fields later and see if I can see a different pattern.

I'd love to be able to suggest direct program changes but I'm only at the start of understanding what is going on....

I'd really like to construct something similar for my Dometic Freshjet AC unit!

**Reply by syssi** [reply]
I'm happy about any feedback. Let's try to improve the implementation if possible.

**Reply by syssi** [reply]
Do you use Discord? Can we have a brief chat here to speed things up?

**Reply by sjoram** [reply]
I haven't used it but just set up an account - soram12345

**Reply by syssi** [reply]
Your Name is Stephen?

**Reply by sjoram** [reply]
Yes
